ipld/amt/src/root.rs

```diff
@@ -1,62 +1,126 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
+use std::marker::PhantomData;
+
 use serde::de::{self, Deserialize};
 use serde::ser::{self, Serialize};
 
 use crate::node::CollapsedNode;
-use crate::{init_sized_vec, Node};
+use crate::{init_sized_vec, Node, DEFAULT_BIT_WIDTH};
+
+#[derive(PartialEq, Debug)]
+pub struct V0;
+#[derive(PartialEq, Debug)]
+pub struct V3;
+
+pub trait Version {
+    const NUMBER: usize;
+}
```

Comment:
I would be careful to fully expose this trait, as now users could implement their own versions and get trapped on unreachable if they ever use it. I would either static assert the version, or make it pub(crate) or something.

Reply:
I have moved it behind a version submodule and pub(crate)'d it.